units/pod_allocator.go

```diff
@@ -0,0 +1,112 @@
+package units
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/evergreen-ci/evergreen"
+	"github.com/evergreen-ci/evergreen/model/pod"
+	"github.com/evergreen-ci/evergreen/model/pod/dispatcher"
+	"github.com/evergreen-ci/evergreen/model/task"
+	"github.com/evergreen-ci/utility"
+	"github.com/mongodb/amboy"
+	"github.com/mongodb/amboy/job"
+	"github.com/mongodb/amboy/registry"
+	"github.com/pkg/errors"
+)
+
+const podAllocatorJobName = "pod-allocator"
+
+func init() {
+	registry.AddJobType(podAllocatorJobName, func() amboy.Job {
+		return makePodAllocatorJob()
+	})
+}
+
+type podAllocatorJob struct {
+	TaskID   string `bson:"task_id" json:"task_id"`
+	job.Base `bson:"job_base" json:"job_base"`
+
+	task *task.Task
+	env  evergreen.Environment
+}
+
+func makePodAllocatorJob() *podAllocatorJob {
+	return &podAllocatorJob{
+		Base: job.Base{
+			JobType: amboy.JobType{
+				Name:    podAllocatorJobName,
+				Version: 0,
+			},
+		},
+	}
+}
+
+func NewPodAllocatorJob(taskID, ts string) amboy.Job {
+	j := makePodAllocatorJob()
+	j.TaskID = taskID
+	j.SetID(fmt.Sprintf("%s.%s.%s", podAllocatorJobName, taskID, ts))
+	j.SetScopes([]string{fmt.Sprintf("%s.%s", podAllocatorJobName, taskID)})
+	j.SetEnqueueAllScopes(true)
+	j.UpdateRetryInfo(amboy.JobRetryOptions{
+		Retryable:   utility.TruePtr(),
+		MaxAttempts: utility.ToIntPtr(10),
+		WaitUntil:   utility.ToTimeDurationPtr(10 * time.Second),
+	})
+
+	return j
+}
+
+func (j *podAllocatorJob) Run(ctx context.Context) {
+	defer j.MarkComplete()
+
+	if err := j.populate(); err != nil {
+		j.AddRetryableError(errors.Wrap(err, "populating job"))
+		return
+	}
+
+	if !j.task.ShouldAllocateContainer() {
+		return
+	}
+
+	intentPod, err := pod.NewTaskIntentPod(pod.TaskIntentPodOptions{
```

Comment:
Does is make sense to do this in the `dispatcher.Allocate` transaction? Or is this already idempotent?

Reply:
It shouldn't matter if this is in the transaction or not since this doesn't perform a DB operation. I tried keeping `Allocate` as simple as possible by only including the necessary transaction + the resulting `PodDispatcher`.